pkg/functions/runner.go

```diff
@@ -0,0 +1,221 @@
+package functions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+)
+
+const (
+	defaultRunHost        = "127.0.0.1" // TODO allow to be altered via a runOpt
+	defaultRunPort        = "8080"
+	defaultRunDialTimeout = 2 * time.Second
+	defaultRunStopTimeout = 10 * time.Second
+	readinessEndpoint     = "/health/readiness"
+
+	// defaultRunTimeout is long to allow for slow-starting functions by default
+	// TODO: allow to be shortened as-needed using a runOption.
+	defaultRunTimeout = 5 * time.Minute
+)
+
+type defaultRunner struct {
+	client *Client
+	out    io.Writer
+	err    io.Writer
+}
+
+func newDefaultRunner(client *Client, out, err io.Writer) *defaultRunner {
+	return &defaultRunner{
+		client: client,
+		out:    out,
+		err:    err,
+	}
+}
+
+func (r *defaultRunner) Run(ctx context.Context, f Function) (job *Job, err error) {
+	var (
+		port    = choosePort(defaultRunHost, defaultRunPort, defaultRunDialTimeout)
+		runFn   func() error
+		verbose = r.client.verbose
+	)
+
+	// Job contains metadata and references for the running function.
+	job, err = NewJob(f, defaultRunHost, port, nil, nil, verbose)
+	if err != nil {
+		return
+	}
+
+	// Scaffold the function such that it can be run.
+	if err = r.client.Scaffold(ctx, f, job.Dir()); err != nil {
+		return
+	}
+
+	// Runner for the Function's runtime.
+	if runFn, err = runFunc(ctx, job); err != nil {
+		return
+	}
+
+	// Run the scaffolded function asynchronously.
+	if err = runFn(); err != nil {
+		return
+	}
+
+	// Wait for it to become available before returning the metadata.
+	err = waitFor(job, defaultRunTimeout)
+	return
+}
+
+// runFunc returns a function which will run the user's Function based on
+// the jobs runtime.
+func runFunc(ctx context.Context, job *Job) (runFn func() error, err error) {
+	runtime := job.Function.Runtime
+	switch runtime {
+	case "go":
+		runFn = func() error { return runGo(ctx, job) }
+	case "python":
+		err = runnerNotImplemented{runtime}
+	case "java":
+		err = runnerNotImplemented{runtime}
+	case "node":
+		err = runnerNotImplemented{runtime}
+	case "rust":
+		err = runnerNotImplemented{runtime}
+	case "":
+		err = fmt.Errorf("runner requires the function have runtime set")
+	default:
+		err = fmt.Errorf("runner does not recognized the %q runtime", runtime)
+	}
+	return
+}
+
+type runnerNotImplemented struct {
+	Runtime string
+}
+
+func (e runnerNotImplemented) Error() string {
+	return fmt.Sprintf("the runtime %q is not supported by the host runner.  Try running containerized.", e.Runtime)
```

Comment:
```suggestion
	return fmt.Sprintf("the runtime %q may only be run containerized.", e.Runtime)
```
Not sure if this is the best wording. I'm just trying to follow convention that errors start with lowercase and do not contain final punctuation. That seems to imply only a single statement.

Reply:
Yes, thank you very much.  Usually when writing these error messages my mind is mostly fully occupied by the logic in development and I tend to not spend enough time ensuring the wording is grammatically correct and informative.